### PR TITLE
Add Eagle3 training for more MLLM model

### PR DESCRIPTION
**Pull request by KerwinKai**
## Motivation

In addition to the previously supported qwen2.5 vl, more multimodal language models have emerged, such as omni, dsvl, internvl, qwen3vl, etc. An abstract solution needs to be designed to easily support the training of draft models of these new models.

## Modifications

1. `QwenVLOnlineEagle3Model` was removed because, for the draft model, only hidden states and the target are needed for SFT fine-tuning; `pixel_values`, `image_grid_thw`, etc., only need to be provided to the target model during the prepare_data stage. Therefore, for the draft model, whether it's LLM or MLLM, it has the same input and output, a structure that can be represented by `OnlineEagle3Model`.
2. The functionality for loading MLLM models is also placed in the `HFEagle3TargetModel` class.
3. Supports the `process_info` function in `preprocessing.py` for different MLLM models to obtain input features for image, video, and audio.

*Testing of this design is still ongoing. I will try to add more support for the aforementioned MLLM models in this PR, and I very much hope to get guidance on whether the changes to this class abstraction are concise enough.*

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by sleepcoo** [comment]
Great job!!!!! This enables us to support all MLLM models. @FrankLeeeee

**Comment by FrankLeeeee** [comment]
What is the current status of this PR now?

**Comment by KerwinKai** [comment]
> What is the current status of this PR now?

Hi, this PR is now ready for review. Meanwhile, under @sleepcoo 's guidance, we are also experimenting with training the draft model for Omni.

**Comment by justadogistaken** [comment]
Hi. Thanks to your work. I've tried training draft for qwen2_5-vl with your work. I found that there is no places to get position_ids from target_model. get_rope_index will give 3D of position_ids, and it works well while trainning qwen vl model. If we don't do it, I guess there will be 2D of position_ids. However `LlamaMutiRotaryEmbedding` needs the 3D of position_ids.

```
position_ids, rope_deltas = self.target_model.model.get_rope_index(
                input_ids,
                image_grid_thw,
                None,
                second_per_grid_ts=None,
                attention_mask=attention_mask_tensor,
            )
```

**Comment by KerwinKai** [comment]
> Hi. Thanks to your work. I've tried training draft for qwen2_5-vl with your work. I found that there is no places to get position_ids from target_model. get_rope_index will give 3D of position_ids, and it works well while trainning qwen vl model. If we don't do it, I guess there will be 2D of position_ids. However `LlamaMutiRotaryEmbedding` needs the 3D of position_ids.
> 
> ```
> position_ids, rope_deltas = self.target_model.model.get_rope_index(
>                 input_ids,
>                 image_grid_thw,
>                 None,
>                 second_per_grid_ts=None,
>                 attention_mask=attention_mask_tensor,
>             )
> ```

Hi, I haven't modified the training logic for Qwen2.5-VL in this PR, but I’ll run tests as soon as possible to identify the issue.

**Comment by FrankLeeeee** [comment]
Does this work for other VLMs which are not Qwen?

**Comment by FrankLeeeee** [comment]
there is conflict with the main branch.

**Comment by KerwinKai** [comment]
> there is conflict with the main branch.

I’ll resolve the merge conflicts as soon as possible and add support for DS-VL and InternVL based on the current architecture.

From the design of the Omni model (see [modeling_qwen3_omni_moe.py#L3976 ](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen3_omni_moe/modeling_qwen3_omni_moe.py#L3976)), our Eagle-based acceleration can only apply to the first-stage thinker model’s generate step, as the talker model already incorporates its own MTP module.

Therefore, to effectively integrate Eagle into Omni’s inference pipeline within SGLang, we likely need to decouple the two-stage forward pass (thinker → talker). Given this, I recommend prioritizing the review of [https://github.com/sgl-project/SpecForge/pull/251 ](https://github.com/sgl-project/SpecForge/pull/251)first.
